esp-hal/src/i2c/master/mod.rs

```diff
@@ -165,6 +165,59 @@ impl BusTimeout {
     }
 }
 
+/// When the FSM remains unchanged for more than the 2^ the given amount of bus
+/// clock cycles a timeout will be triggered.
+///
+/// The default value is 0x10
+#[instability::unstable]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[cfg(not(any(esp32, esp32s2)))]
+pub struct FsmTimeout {
+    value: u8,
+}
+
+#[cfg(not(any(esp32, esp32s2)))]
+impl FsmTimeout {
+    const FSM_TIMEOUT_MAX: u8 = 23;
+
+    const FSM_TIMEOUT_DEFAULT: u8 = 0x10;
+
+    /// Creates a new timeout.
+    ///
+    /// The meaning of the value and the allowed range of values is different
+    /// for different chips.
+    pub const fn new_const<const VALUE: u8>() -> Self {
+        const {
+            core::assert!(VALUE <= Self::FSM_TIMEOUT_MAX, "Invalid timeout value");
+        }
+        Self { value: VALUE }
+    }
+
+    /// Creates a new timeout.
+    ///
+    /// The meaning of the value and the allowed range of values is different
+    /// for different chips.
+    pub fn new(value: u8) -> Result<Self, ConfigError> {
```

Comment:
Can't we just mark this is `const`? It looks to me like the code should be able to compile at const time, then there is no need for the `new_const` method?

Reply:
Initially it was `const` but the check still ended up as a runtime check even if the argument is constant: https://github.com/esp-rs/esp-hal/pull/3333#discussion_r2029031499

Reply:
Sorry I missed that discussion :sweat_smile:, thanks!